source/fundamentals/class-mapping.txt

```diff
@@ -0,0 +1,328 @@
+element that's added to a document to identify which class the document was
+serialized from.
+
+You can specify a discriminator using the ``BsonDiscriminator`` attribute as
+follows:
+
+.. code-block:: csharp
+
+   [BsonDiscriminator("personClass")]
+   public class Person 
+   {
+       public string Name { get; set; }
+       public int Age { get; set; }
+       public List<string> Hobbies {get; set;}
+   }
+
+You can also specify a discriminator when registering a class map as follows:
+
+.. code-block:: csharp
+
+   BsonClassMap.RegisterClassMap<Person>(classmap => 
+   {
+       classmap.AutoMap();
+       classmap.SetDiscriminator("personClass");
+   });
+
+In BSON, discriminators have the field name ``_t``. 
+
+The following example shows how a document from the ``Person`` class with the
+"personClass" discriminator appears in the collection after serialization:
+
+.. code-block:: json
+
+   { "_id": "...", "_t": "personClass", "name": "...", "age": "...", "hobbies": [...]}
+
+.. TODO: Link to page on polymorphism/discriminators
+
+Mapping with Constructors
+-------------------------
+
+By default, the {+driver-short+} can automatically map a class only if the class has
+a constructor with no arguments. If you want the driver to use a constructor that accepts
+one or more arguments, you can add the ``BsonConstructor`` attribute to the constructor.
+In this case, the driver parses the expression tree to determine how to map the   
+constructor arguments to class properties or fields.
+
+When the driver creates a class map for the following ``Person`` class, it will use the 
+constructor marked with the ``BsonConstructor`` attribute. The ``name`` argument will
+map to the ``Name`` property and the ``age`` argument will map to the ``Age`` property.
+
+.. code-block:: csharp
+   :emphasize-lines: 7
+
+   public class Person 
+   {
+       public string Name { get; set; }
+       public int Age { get; set; }
+       public List<string> Hobbies {get; set;}
+
+       [BsonConstructor]
+       public Person(string name, string age)
+       {
+           Name = name;
+           Age = age;
+       }
+   }
+
+.. tip:: Multiple ``BsonConstructor`` attributes
+
+   If there is more than one constructor with the ``BsonConstructor``
+   attribute, the driver uses the constructor that has the most
+   parameters with matching fields in the document.
+
+You can also specify the constructor to use when registering your class map:
+
+.. code-block:: csharp
+   :emphasize-lines: 17
+
+   public class Person 
+   {
+       public string Name { get; set; }
+       public int Age { get; set; }
+       public List<string> Hobbies {get; set;}
+
+       public Person(string name, string age)
+       {
+           Name = name;
+           Age = age;
+       }
+   }
+
+   BsonClassMap.RegisterClassMap<Person>(classmap =>
+   {
+       classmap.AutoMap();
+       classmap.MapCreator(p => new Person(p.Name, p.Age));
```

Comment:
I don't think `MapCreator` used like this will work.

Have you verified it?

Normally you would pass `MapCreator` a static factory method.

Reply:
There is also a `MapConstructor` method but it requires passing it a `ConstructorInfo` value which you would have to get using reflection.

Reply:
I took this code block from the current documentation here:
https://mongodb.github.io/mongo-csharp-driver/2.17/reference/bson/mapping/#creation-customization
(just changing the property names to match the other examples on this page). Is this information incorrect, or did I perhaps mistranslate it?

Reply:
I'll have to write a test to find out. I'll get back to you.